Add support for traces

nmn opened the pull request:
AtomLinter supports arrays of traces in every error object. So now there is a correct way of dealing with flow errors that span multiple files.

nmote commented on pkg/nuclide/flow/lib/FlowLinter.js:
```diff
@@ -71,12 +74,10 @@ function processDiagnostics(diagnostics: Array<Object>, targetFile: string) {
-  return messages.map(flowMessageToLinterMessage);
+  return diagnostics
+    .map( (diagnostic) => diagnostic['message'] )
+    .filter(hasMessageWithPath);
```
I'm getting a parse error here, due to this extraneous semicolon. Did you actually try running this?

nmn commented:
@nmote I actually did a little code reformatting to match the style after running it. May have made a mistake. Let me see.

nmn commented:
Should be fixed now. Sorry again.

nmote commented:
Thanks, checking this out.

nmote commented on pkg/nuclide/flow/lib/FlowLinter.js:
```diff
@@ -30,39 +30,42 @@ var {JS_GRAMMARS} = require('./constants.js');
-  };
+// A trace object is very similar to an error object.
+function flowMessageToTrace(message){
+  return { 
```
Please remove the trailing whitespace

nmote commented on pkg/nuclide/flow/lib/FlowLinter.js:
```diff
@@ -30,39 +30,42 @@ var {JS_GRAMMARS} = require('./constants.js');
+    text: message.descr,
+    filePath: message.path,
+    range: extractRange(message)
+  }
```
Please add a semicolon here

nmote commented on pkg/nuclide/flow/lib/FlowLinter.js:
```diff
@@ -30,39 +30,42 @@ var {JS_GRAMMARS} = require('./constants.js');
-  message['descr'] = messages.map((msg)=>msg['descr']).join(' ');
-  return message;
+function flowMessageToLinterMessage(arr) {
+  var message = arr[0]
```
semicolon

nmote commented on pkg/nuclide/flow/lib/FlowLinter.js:
```diff
@@ -30,39 +30,42 @@ var {JS_GRAMMARS} = require('./constants.js');
+function flowMessageToLinterMessage(arr) {
+  var message = arr[0]
+
+  var obj = { 
```
trailing whitespace

nmote commented on pkg/nuclide/flow/lib/FlowLinter.js:
```diff
@@ -30,39 +30,42 @@ var {JS_GRAMMARS} = require('./constants.js');
+
+  var obj = { 
+    type: message.level,
+    text: arr.map(obj => obj.descr).join(' '),
```
Since `obj` is already defined in this scope, please use a different variable name here to avoid shadowing.

nmote commented on pkg/nuclide/flow/lib/FlowLinter.js:
```diff
@@ -30,39 +30,42 @@ var {JS_GRAMMARS} = require('./constants.js');
+    text: arr.map(obj => obj.descr).join(' '),
+    filePath: message.path,
+    range: extractRange(message)
+  }
```
semicolon

nmote commented on pkg/nuclide/flow/lib/FlowLinter.js:
```diff
@@ -30,39 +30,42 @@ var {JS_GRAMMARS} = require('./constants.js');
+
+  // When the message is an array with multiple elements
+  // The second element onwards make the trace for the error.
+  if(arr.length > 1){
```
nit: there ought to be a space after the `if`

nmote commented on pkg/nuclide/flow/lib/FlowLinter.js:
```diff
@@ -30,39 +30,42 @@ var {JS_GRAMMARS} = require('./constants.js');
+  // When the message is an array with multiple elements
+  // The second element onwards make the trace for the error.
+  if(arr.length > 1){
+    obj.trace = arr.slice(1).map(flowMessageToTrace)
```
semicolon

nmote commented on pkg/nuclide/flow/lib/FlowLinter.js:
```diff
@@ -30,39 +30,42 @@ var {JS_GRAMMARS} = require('./constants.js');
+    obj.trace = arr.slice(1).map(flowMessageToTrace)
+  }
+
+  return obj
```
semicolon

nmote commented on pkg/nuclide/flow/lib/FlowLinter.js:
```diff
@@ -71,12 +74,10 @@ function processDiagnostics(diagnostics: Array<Object>, targetFile: string) {
-  return messages.map(flowMessageToLinterMessage);
+  return diagnostics
+    .map( (diagnostic) => diagnostic['message'] )
+    .filter(hasMessageWithPath)
```
Each message is an array of objects. `hasMessageWithPath` expects an object with a `path` property. So, it always returns false and all linter messages are filtered out.

nmote commented:
Additionally, `apm test` fails. Please make sure all tests pass and make sure that Flow errors actually appear in the editor.

nmn commented:
Just pushed more changes. Here's some info that might help you avoid wasting time:
1. I added an eslintrc file to help me catch styling problems, and added the missing semi-colons, removed trailing spaces, and add dangling commas
1a. I made dangling commas mandatory, which meant I had to add a few dangling commas where they didn't exist before.
2. I Used setup script and tested that the errors work correctly in Atom
3. I Updated the tests to test the code correctly -
3a. Errors are not always of the type 'Error', they are either 'error' or 'warning' which is taken directly from the flow diagnostic. I have tested that the lower-case strings work fine with AtomLinter
3b. I changed the last test to check for the trace array being present on the object.

Other stuff:
1. I added a couple of flow type definitions in the file to be a stricter type check for diagnostic objects
2. I would like to commit the `.eslintrc` file to help future contributors, let me know if that is OK.

nmn commented:
Just made another commit to use square bracket notation instead of dot notation. This is more consistent with the current code style.

nmote commented:
Thanks for the updates. I probably won't have a chance to review this today, so it will likely wait until Monday.

Thanks also for setting up an `.eslintrc`. We actually already have linting set up internally, it just hasn't made it to the GitHub repository. It's probably trivial but it just has to be done. I'll do it next week.

nmote commented:
This looks good to me! Thanks!

I have to get a second opinion before merging. You should hear back by Monday or Tuesday.

nmn commented:
Thanks, waiting for this and `on-the-fly linting` to land soon.

nmote commented:
@nmn: https://github.com/facebook/nuclide/commit/11401bc407d6d089f17577c8a707d4ca16e67929, which enables on-the-fly linting, is already landed.

nmn commented:
@nmote Awesome. That's great. Will wait for this PR to be merged then. And all this being released.

nmote commented:
We merged this with a couple more minor style changes. It should be out soon! Thanks for contributing!

nmn commented:
Any update on a new Nuclide release?

bolinfest commented:
I'll aim to push one out this weekend. I just exported a bunch of new code
to GitHub.

On Wed, Jul 29, 2015 at 3:02 PM, Naman Goel <notifications@github.com>
wrote:

> Any update on a new Nuclide release?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/facebook/nuclide/pull/139#issuecomment-126109588>.
>

bolinfest commented:
@nmn Nuclide v0.0.25 is out!
